### PR TITLE
More miscellaneous fixes and tidy‐ups

### PR DESCRIPTION
**Pull request by josephsutton1**

### CODE REVIEW

**Review by abartlet — APPROVED** [review]
These look really good.  @nicowilliams can we/I merge these please?

**Comment by nicowilliams** [comment]
> These look really good.  @nicowilliams can we/I merge these please?

I'll look in a bit.  Sorry I've been AWOL.

File: kdc/fast.c [hunk 1]
```diff
@@ -373,8 +373,8 @@ _kdc_fast_mk_error(astgs_request_t r,
 			    NULL,
 			    error_client,
 			    error_server,
-			    NULL,
-			    NULL,
+			    csec,
```
Comment (nicowilliams) [inline kdc/fast.c:376]: Did we have a bug here?

File: kdc/fast.c [hunk 2]
```diff
@@ -549,9 +552,14 @@ fast_unwrap_request(astgs_request_t r,
...
 
+	if (ap_req.ticket.enc_part.kvno != NULL) {
+	    kvno = *ap_req.ticket.enc_part.kvno;
+	    kvno_ptr = &kvno;
+	}
+
 	ret = _kdc_db_fetch(r->context, r->config, armor_server_principal,
 			    HDB_F_GET_KRBTGT | HDB_F_DELAY_NEW_KEYS,
-			    (krb5uint32 *)ap_req.ticket.enc_part.kvno,
+			    kvno_ptr,
```
Comment (nicowilliams) [inline kdc/fast.c:562]: Nice.

File: lib/hdb/hdb-ldap.c [hunk 3]
```diff
@@ -357,7 +357,7 @@ LDAP_get_generalized_time_value(HDB * db, LDAPMessage * entry,
 				const char *attribute, KerberosTime * kt)
 {
     char *tmp, *gentime;
-    struct tm tm;
+    struct tm tm = {};
```
Comment (nicowilliams) [inline lib/hdb/hdb-ldap.c:360]: Did a compiler complain about this?

File: lib/hx509/cms.c [hunk 4]
```diff
@@ -935,7 +935,7 @@ hx509_cms_verify_signed_ext(hx509_context context,
 	if (signer_info->signature.length == 0) {
 	    ret = HX509_CMS_MISSING_SIGNER_DATA;
 	    hx509_set_error_string(context, 0, ret,
-				   "SignerInfo %d in SignedData "
+				   "SignerInfo %zu in SignedData "
```
Comment (nicowilliams) [inline lib/hx509/cms.c:938]: Do old MSVC compilers and runtimes support this?

File: lib/hx509/hxtool.c [hunk 5]
```diff
@@ -2898,13 +2898,14 @@ acert1_kus(struct acert_options *opt,
...
 {
-    struct tm at_tm;
+    struct tm at_tm = {};
     char *rest;
     int at_s;
 
     if ((rest = strptime(s, "%Y-%m-%dT%H:%M:%S", &at_tm)) != NULL &&
         rest[0] == '\0')
         return mktime(&at_tm);
+    at_tm = (struct tm) {};
```
Comment (nicowilliams) [inline lib/hx509/hxtool.c:2908]: Does MSVC support this?

File: kdc/fast.c (hunk 1 above)
Comment (josephsutton1) [reply kdc/fast.c:376]: Possibly? Starting with commit 9b55215a2aece1268c9c2cf84ebc8e82cb72625a (“kdc: sync KDC FAST with Heimdal-597.121.1”) we’ve omitted `csec` and `cusec` completely when FAST is in use. Including them in the inner error is more consistent with the handling of `r->e_text`, `error_client`, and `error_server`, which are included in the inner error but omitted from the outer.

File: lib/hdb/hdb-ldap.c (hunk 3 above)
Comment (josephsutton1) [reply lib/hdb/hdb-ldap.c:360]: No, just something I noticed when I was debugging an unrelated issue.

File: lib/hx509/cms.c (hunk 4 above)
Comment (josephsutton1) [reply lib/hx509/cms.c:938]: I don’t have an old MSVC compiler to hand, sorry.

Comment (josephsutton1) [reply lib/hx509/cms.c:938]: But I see `%zu` is used in several places in Heimdal already.

File: lib/hx509/hxtool.c (hunk 5 above)
Comment (nicowilliams) [reply lib/hx509/hxtool.c:2908]: @jaltman says "no".  Can you use `memset()` instead?

File: lib/hx509/cms.c (hunk 4 above)
Comment (nicowilliams) [reply lib/hx509/cms.c:938]: Ah, @jaltman tells me that roken supports `%zu`.

File: lib/hx509/hxtool.c [hunk 6]
```diff
@@ -2898,13 +2898,14 @@ acert1_kus(struct acert_options *opt,
 static time_t
 ptime(const char *s)
 {
-    struct tm at_tm;
+    struct tm at_tm = {};
```
Comment (nicowilliams) [inline lib/hx509/hxtool.c:2901]: Use `memset()` :(

File: lib/hx509/hxtool.c (hunk 5 above)
Comment (josephsutton1) [reply lib/hx509/hxtool.c]: OK.

**Comment by nicowilliams** [comment]
LGTM.  I'll merge after checks pass.  Thanks!

**Comment by abartlet** [comment]
After one restart (iprop) checks pass!

**Comment by nicowilliams** [comment]
Thanks!

**Pull request merged**